### ISSUE TITLE
[Question] How does target binary exactly run in guest VM?

### ISSUE THREAD

**Issue by Hanseltu** [S2E/s2e-env#460]
Hello S2E developers,

Thanks for your nice work on building a cool S2E! I am currently learning symbolic execution techniques as well as S2E and I hope this is the right place to raise a question. 

My core question is about how a target binary runs in a guest VM launched by S2E. From S2E documents, I have known that the guest images can be modified by users or added with new software (based on the introduction in [guest-image](https://github.com/S2E/guest-images)), so there will be a lot of applications inside a guest image when we analyze a target binary. Here, my question is that **when the target binary runs in guest VM, are any other applications still actively run**? For example, when using S2E to analyze [Microsoft Office](http://s2e.systems/docs/Tutorials/MSOffice/index.html), does this mean only the current analyzed binary. i.e., `winword.exe` in `windows-7sp1pro-i386/office2016` image, is actively running, while other applications, i.e., the desktop application or some required service are sleeping?

To find the answer, I have read the guideline for [Customizing stock VM images](http://s2e.systems/docs/ImageInstallation.html#id1). In the final part, some suggestions to disable unnecessary tasks are proposed to make a fast program analysis in S2E. So, does this mean there are many applications along with target binary are executing in guest VM although these activities may slow down the analysis process? If the answer is yes, how does S2E handle the scheduling between different applications in guest VM? Assuming some target binaries may have interactions with other applications, how does S2E handle such a scenario and run those programs correctly?

I have also tried to install S2E manually in my machine (ubuntu 20.04) to try to understand the S2E's behaviors deeply, however, I am sorry I am stuck at the beginning stage when executing `s2e build`, based on the setup [tutorial](http://s2e.systems/docs/s2e-env.html). I encountered a build error (`/build/llvm-10.0.0.src/utils/benchmark/src/benchmark_register.h:17:18: error: default initialization of an object of const type 'const long' static const t max = std::numeric_limits<t>::max();`) when building LLVM, I am not sure where the problem comes from (probably is the compatibility issue of LLVM compiler versions in my computer or in S2E). 

In summary, my question is that when a target binary is executed in guest VM, are other applications in the VM actively running or just keeping silent to stop running until the analysis of the target program is done? If there can be a lot of programs running in guest VM simultaneously, how does S2E handle the scheduling issue between different programs? If it's true that only the target binary can be executed in the guest VM, how did you configure the building of the guest image to enable such a running environment?

I hope I made a clear presentation of my question and looking forward to your response! Thank you so much!

Best regards,
Haoxin

**Comment by vitalych** [comment]
Hi!

S2E runs an entire VM. It's a normal VM, that runs normal unmodified OS and software. Actually, S2E doesn't really care what it runs. The symbolic execution engine sees a stream of x86 instructions. Likewise, the OS doesn't really know that it runs in S2E. All programs will run like on a normal computer (with a somewhat large performance overhead though).

S2E comes with a large set of plugins that are OS-aware. They let you do things like record execution traces, get code coverage, etc. These plugins do need to be aware of the OS that is running in the guest in order to make sense of the instructions that are being executed. S2E comes with a set of guest utilities and drivers that help its plugins make sense of where the programs are. You can see these guest tools in the boostrap.sh file of an S2E project.

These plugins and guest tools are entirely optional. You will still be able to run symbolic execution without any of them. In order to achieve this, S2E extends the x86 instruction set with a special instruction that injects symbolic data in the specified memory location. That instruction also lets the guest communicate with plugins. Everything else builds on top of that.

Regarding your issue, it may be an issue with the compiler. I'll try to upgrade LLVM soon.

**Comment by Hanseltu** [comment]
Hi Vitaly,

Thanks for your super speedy and insightful response!

Your answer makes sense to me. May I ask further questions for you?

Regarding the execution without levering plugins and guest tools, are those symbolic data injected by extended instructions in S2E always be randomly performed or based on some pre-defined patterns or can be defined anywhere by users like KLEE? The fact is that there may be many buffers or variables existing in the target program, and how does S2E select the interesting/promising points that are made to be symbolic? (I am sorry I am not much familiar with S2E's implementation)

Apart from the above concern, another confusion in my mind is how does S2E analyze an entire Linux kernel, i.e., does it support exploring multiple states in the kernel when the kernel invokes symbolic data in conditions? In the example presented in your paper (see below, `sysFn` function in the kernel), it seems only one single thread of concrete execution is performed in this sample example. That means, when encountering symbolic data in kernel execution, S2E always concretizes the symbolic variables and performs concrete execution. Am I right here? Outside this paper, there are rare studies that are analyzing entire kernel behavior using S2E (only for drivers, e.g., [SymDrive](https://research.cs.wisc.edu/sonar/projects/symdrive/downloads.shtml)). I know the job on symbolic execution of the entire kernel could be really hard, but how is the difficulty of implementing S2E to conduct symbolic execution on the entire kernel? Are any other major challenges for S2E to analyze the entire kernel behavior?

<img width="583" alt="se2" src="https://user-images.githubusercontent.com/19731522/156005262-eb5a78aa-1f8d-431f-af95-a8592af9852d.png">

Thank you so much again for your helpful comments and suggestions！

Best,
Haoxin

**Comment by Hanseltu** [comment]
OK! Thanks and let me try then.
